storage/badger/approvals.go

```diff
@@ -0,0 +1,137 @@
+		return operation.SkipDuplicates(operation.InsertResultApproval(approval))
+	}
+
+	retrieve := func(key interface{}) func(tx *badger.Txn) (interface{}, error) {
+		approvalID := key.(flow.Identifier)
+		var approval flow.ResultApproval
+		return func(tx *badger.Txn) (interface{}, error) {
+			err := operation.RetrieveResultApproval(approvalID, &approval)(tx)
+			return &approval, err
+		}
+	}
+
+	res := &ResultApprovals{
+		db: db,
+		cache: newCache(collector,
+			withLimit(flow.DefaultTransactionExpiry+100),
+			withStore(store),
+			withRetrieve(retrieve),
+			withResource(metrics.ResourceResult)),
+	}
+
+	return res
+}
+
+func (r *ResultApprovals) store(approval *flow.ResultApproval) func(*badger.Txn) error {
+	return r.cache.Put(approval.ID(), approval)
+}
+
+func (r *ResultApprovals) byID(approvalID flow.Identifier) func(*badger.Txn) (*flow.ResultApproval, error) {
+	return func(tx *badger.Txn) (*flow.ResultApproval, error) {
+		val, err := r.cache.Get(approvalID)(tx)
+		if err != nil {
+			return nil, err
+		}
+		return val.(*flow.ResultApproval), nil
+	}
+}
+
+func (r *ResultApprovals) byChunk(resultID flow.Identifier, chunkIndex uint64) func(*badger.Txn) (*flow.ResultApproval, error) {
+	return func(tx *badger.Txn) (*flow.ResultApproval, error) {
+		var approvalID flow.Identifier
+		err := operation.LookupResultApproval(resultID, chunkIndex, &approvalID)(tx)
+		if err != nil {
+			return nil, fmt.Errorf("could not lookup result approval ID: %w", err)
+		}
+		return r.byID(approvalID)(tx)
+	}
+}
+
+func (r *ResultApprovals) index(resultID flow.Identifier, chunkIndex uint64, approvalID flow.Identifier) func(*badger.Txn) error {
+	return func(tx *badger.Txn) error {
+		err := operation.IndexResultApproval(resultID, chunkIndex, approvalID)(tx)
+		if err == nil {
+			return nil
+		}
+
+		if !errors.Is(err, storage.ErrAlreadyExists) {
+			return err
+		}
+
+		// When trying to index an approval for a result, and there is already
+		// an approval for the result, double check if the indexed approval is
+		// the same.
+		// We don't allow indexing multiple approvals per chunk because the
+		// store is only used within Verification nodes, and it is impossible
+		// for a Verification node to compute different approvals for the same
+		// chunk.
+		var storedApprovalID flow.Identifier
+		err = operation.LookupResultApproval(resultID, chunkIndex, &storedApprovalID)(tx)
+		if err != nil {
+			return fmt.Errorf("there is an approval stored already, but cannot retrieve it: %w", err)
+		}
+
+		if storedApprovalID != approvalID {
+			return fmt.Errorf("attempting to store conflicting approval (result: %v, chunk index: %d): storing: %v, stored: %v. %w",
+				resultID, chunkIndex, approvalID, storedApprovalID, storage.ErrDataMismatch)
+		}
+
+		return nil
+	}
+}
+
+// Store stores a ResultApproval
+func (r *ResultApprovals) Store(approval *flow.ResultApproval) error {
+	return operation.RetryOnConflict(r.db.Update, r.store(approval))
```

Comment:
I think that `Store` doesn't need to check that the value is consistent because it uses the `ID` of the item as key. 
It doesnt need to account for a situation where two inconsistent values are colliding with the same key. So I think the `store` method is idempotent too.

The situation is different with `index` because the key (result ID, chunk index) is not necessarily unique per chunk.. It is possible for two different verifiers to compute different chunks for the same (resultID, chunk index).

Reply:
💡